ip_k66f board refactor

lmajewski opened the pull request:
Enhancements for ip_k66f board (embOS/IP switch):
- Enable SPI and ETH support (also include proper pinmux setup)
- Enable support for RTT console communication.
- Add support for using external clock as a RMII reference
- Add support for K66F running in High Speed RUN mode (not available on K64F) 
- The ip_k66f board now runs with 180MHz (instead of 120MHz).

Commit:
board: ip_k66f: Add support for RTT based console

The ip_k66f board (embOS/IP switch) from Segger doesn't have the serial
console pins connected to J-Link OB. As a result one needs to use RTT
to get the serial console.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dts: ip_k66f: Enable eth on the ip_k66f board

Enable ETH in device tree for ip_k66f.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dts: ip_k66f: Enable spi1 on the ip_k66f board

Enable SPI1 in device tree for ip_k66f.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
ip_k66f: pinmux: eth: Add ETH support

Modify pinmux to properly configure ETH pins.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
ip_k66f: pinmux: Add pins configuration for SPI1

Modify pinmux to properly configure SPI1 pins.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
defconfig: Enable support for SPI on ip_k66f

Modify ip_k66f configuration files to support SPI.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
Kconfig: Enable support for ETH on ip_k66f

Modify ip_k66f configuration files to support ETH.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
defconfig: ip_k66f: Set proper CONFIG_OSC_XTAL0_FREQ to 12MHz

The CONFIG_OSC_XTAL0_FREQ shall be defined only once, so remove
it from board specific defconfig.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
clk: ip_k66f: Fix ip_k66f to use 180MHz as a system clock

The ip_k66f board can run with clock frequency of 180 MHz. This requires
switching it to high speed mode (HSRUN).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
clk: k6x: Modify clock_init to enable external RMII clock source

This change enables the RMII external clock source at the clock
initialization function.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
ip_k66f: pinmux: Add support for PTE26 to work as clock source for RMII

Modify pinmux to properly configure ETH's clock source pin for RMII.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
defconfig: Increase the size of RTT console "UP" buffer

This change allows usage of larger RTT "UP" buffers from target to host.
It is necessary to allow correct execution of ping command:

net ping 192.168.0.1

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
clk: eth: Add ETH_MCUX_RMII_EXT_CLK option to enable external RMII clock

This option will configure MCUX block (by setting RMIISRC [19] bit to 1
in SIM_SOPT2 register) to use external clock source for RMII from
ENET_1588_CLKIN).

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
defconfig: Enable RMII_EXT_CLK on ip_k66f board

The embOS/IP switch board (ip_k66f) uses external clock (from KSZ8794
SoC) as a reference for its RMII.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
clk: k6x: Enable High Speed RUN (HSRUN) mode for K66F SOC

The NXP's Kinetics K66F is able to run with max frequency of 180MHz.
To achieve this goal the SMC's PMPROT and PMCTRL registers need to be
adjusted.

On the contrary the K64F doesn't support HSRUN run mode.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
k6x: Kconfig: Add CONFIG_K6X_HSRUN option for K66F

This option enables support for High Speed RUN operation mode for
K66F. The K64F SoC doesn't support this mode.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

galak commented on boards/arm/ip_k66f/Kconfig.defconfig:
```diff
@@ -9,17 +9,39 @@ config BOARD
 	default 1
 
+if SPI
+config SPI_1
```
you can remove this bit about SPI_1 as that is driven by the devicetree now.

galak commented on boards/arm/ip_k66f/pinmux.c:
```diff
@@ -17,10 +18,47 @@ static int ip_k66f_pinmux_init(struct device *dev)
 	pinmux_pin_set(porta, 8, PORT_PCR_MUX(kPORT_MuxAsGpio));
 	pinmux_pin_set(porta, 10, PORT_PCR_MUX(kPORT_MuxAsGpio));
 
+#if CONFIG_ETH_MCUX_0
```
Use nodelabel, see `boards/arm/frdm_k64f/pinmux.c` for example.

galak commented on boards/arm/ip_k66f/pinmux.c:
```diff
@@ -17,10 +18,47 @@ static int ip_k66f_pinmux_init(struct device *dev)
+	pinmux_pin_set(porte, 26, PORT_PCR_MUX(kPORT_MuxAlt2));
+#endif
+
+#if CONFIG_SPI_1
```
use nodelabel here as well.

Commit:
ip_k66f: Kconfig: Enable PINMUX_MCUX_PORTB only when "spi1" label defined

The CONFIG_SPI_1 shall be replaced with condition based on presence of
device tree 'spi1' node label.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

Commit:
dts: ip_k66f: Enable gpio{abe} on the ip_k66f board

Enable GPIOA, GPIOB and GPIOE in the device tree for ip_k66f.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

jfischer-phytec-iot commented on boards/arm/ip_k66f/ip_k66f.dts:
```diff
@@ -93,3 +94,10 @@
+&spi1 {
+	status = "okay";
+
+	cs-gpios = <&gpiob 10 GPIO_ACTIVE_LOW>;
```
AFAIK spi_mcux_dspi supports PCS lines, I would not use gpio for CS on this platform.

lmajewski replied:
I've checked at least two times if I could use PCS. The PCS was not working and as fair as I understand the issue was with the SPI driver for K6x devices not driving this line at all.

I can check third time if the PCS is supported by NXP's SPI driver for K6x

MaureenHelm replied:
The driver definitely supports PCS; we use it on the frdm_k64f board.

jfischer-phytec-iot commented on boards/arm/ip_k66f/pinmux.c:
```diff
@@ -32,6 +37,15 @@ static int ip_k66f_pinmux_init(struct device *dev)
+
+#if DT_HAS_NODE_STATUS_OKAY(DT_NODELABEL(spi1))
+	/* SPI1 CS0, SCK, SOUT, SIN - Control of KSZ8794 */
+	pinmux_pin_set(portb, 10, PORT_PCR_MUX(kPORT_MuxAsGpio));
```
You could mux it to PCS, probably it is kPORT_MuxAlt2 (SPI1_PCS0), see also https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/frdm_k64f/pinmux.c#L106

jfischer-phytec-iot commented on boards/arm/ip_k66f/pinmux.c:
```diff
@@ -55,6 +56,11 @@ static int ip_k66f_pinmux_init(struct device *dev)
 	pinmux_pin_set(portb, 17, PORT_PCR_MUX(kPORT_MuxAlt2));
 #endif
 
+	struct device *gpioa = device_get_binding(DT_ALIAS_GPIO_A_LABEL);
```
I do not think pinmux.c is the right place for that but probably board.c

lmajewski replied:
Ok. I will introduce board.c and add there the RESET gpio setup pins.

jfischer-phytec-iot commented on boards/arm/ip_k66f/Kconfig.defconfig:
```diff
@@ -20,6 +20,19 @@ config MCG_VDIV0
+if NETWORKING
+
+config NET_L2_ETHERNET
+	default y if !MODEM
```
Do you have MODEM on this board?

lmajewski replied:
No, there is no modem. I think that it was a leftover from frdm_k64f board. I will remove the !MODEM

MaureenHelm reviewed (commented):
Agree with the suggestions from @jfischer-phytec-iot, but otherwise it looks good

henrikbrixandersen commented on boards/arm/ip_k66f/Kconfig.board:
```diff
@@ -5,3 +5,13 @@ config BOARD_IP_K66F
+	int
+	default 47
+	depends on BOARD_IP_K66F
+	depends on GPIO
```
Having this depend on GPIO will probably not do what you want here. If GPIO is disabled, `CONFIG_BOARD_INIT_PRIORITY` will be undefined yet `board.c` will still be compiled?

lmajewski replied:
You are correct - the dependency on GPIO shall be removed.

henrikbrixandersen commented on boards/arm/ip_k66f/board.c:
```diff
@@ -0,0 +1,23 @@
+		device_get_binding(DT_LABEL(DT_NODELABEL(gpioa)));
+
+	/* Set !ETH_RST */
+	gpio_pin_configure(gpioa, 24, GPIO_OUTPUT_HIGH);
```
Another way of handling this could be to have a reset GPIO for the PHY and have the ethernet driver bring the PHY out of reset before use. With the current approach, the PHY will be taken out of reset regardless of whether ethernet is to be used (I know, on this board you probably have more use-cases with ethernet enabled than without, but still).

bwasim replied:
I would also recommend to get this information from the device tree file and handle reset in the Ethernet driver itself.. This would increase the portability of the driver somewhat (next time no one will have to write a board level device tree file for PHY reset)..

lmajewski replied:
Could you be more specific about the "Ethernet" driver? Would you recommend adding the generic PHY reset code into eth_mcux.c (which is NXP's Kinetics specific) or maybe to ethernet.c's ethernet_init() function (which looks to be more generic)?

lmajewski replied:
I will drop this code - i.e. the board.c file from this patch set. Instead I will introduce new PR and patches to address this topic (as it seems to be more generic than fixes for this particular board).

lmajewski replied:
But I would still like to hear your recommendation if this reset code shall be put into eth_mcux.c or ethernet.c ?

jukkar replied:
The subsys/net/l2/ethernet/ethernet.c is for generic, hw agnostic L2 code. This means that it is not suppose to have any gpio configure calls. So all of this kind of code should stay inside drivers/ethernet directory. You can create some common file there if you think more than one driver would benefit from such calls.

bwasim replied:
We should do something similar to how the linux kernel handles ethernet phy, and ethernet phy resets. 
Please see https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/net/ethernet-phy.yaml#L141 which defines the "reset-gpios" property for Ethernet PHYs. 

As this is generic across all Ethernet PHYs, it should be placed in a common file and there should be a global function (accessible by all Ethernet drivers) to optionally call up this function at boot up. 

You can take https://patchwork.kernel.org/patch/7871481/ as reference which adds support for "reset-gpios" binding in the cadence gem driver (AKA macb)..

lmajewski replied:
Ok, thanks for the input. It turns out that HW on the ip_k66f board resets the PHY, so there is no urgent need for such code.
I've already dropped the board.c from this patch set and will introduce it when needed.

henrikbrixandersen commented on boards/arm/ip_k66f/ip_k66f_defconfig:
```diff
@@ -8,9 +8,18 @@ CONFIG_GPIO=y
+CONFIG_K6X_FLASH_CLOCK_DIVIDER=7
 CONFIG_USE_SEGGER_RTT=y
 
+# Add RTT console support
+CONFIG_CONSOLE=y
+CONFIG_RTT_CONSOLE=y
+CONFIG_SEGGER_RTT_BUFFER_SIZE_DOWN=64
+
```
Perhaps mention in the board documentation that this board, unlike most if not all others in-tree, default to using Segger RTT for console?

lmajewski replied:
I will extend the documentation entry.

henrikbrixandersen commented on soc/arm/nxp_kinetis/k6x/Kconfig.soc:
```diff
@@ -117,4 +117,12 @@ config K6X_FLASH_CLOCK_DIVIDER
+	default y
+	help
+	  This options enables support for High Speed RUN mode on K66F SoC. It is
+	  not available on K64F.
```
Do we need to mention K64F here? The option already depends on `SOC_MK66F18`, so it will never appear in the Kconfig menus for a K64F based board.

lmajewski replied:
Ok,

Commit:
doc: Update ip_k66f documentation entry about serial console

This patch adds information regarding usage of serial console on
the ip_k66f board, as it only can use Segger's RTT console for
communication.

Signed-off-by: Lukasz Majewski <lukma@denx.de>

MaureenHelm commented:
will you please rebase to the latest master?
